File: solutions/python/binary-search/2/binary_search.py
"""This module provides a function to find the index of a value in a list."""


def find(search_list: list[int], value: int) -> int:
    """Perform binary search to find the index of a value in a sorted list.

    Args:
        search_list (list[int]): The list to search in.
        value (int): The value to search for.

    Returns:
        int: The index of the value in the list.

    Raises:
        ValueError: If the value is not found in the list.
    """

    def binary_search(arr: list[int], low: int, high: int, value_: int) -> int:
        """Recursive binary search algorithm.

        Args:
            arr (list[int]): The list to search in.
            low (int): The starting index of the search range.
            high (int): The ending index of the search range.
            value_ (int): The value to search for.

        Returns:
            int: The index of the value in the list.

        Raises:
            ValueError: If the value is not found in the list.
        """
        if high >= low:
            mid = (high + low) // 2
            if arr[mid] == value_:
                return mid
            if arr[mid] > value_:
                return binary_search(arr, low, mid - 1, value_)
            return binary_search(arr, mid + 1, high, value_)
        raise ValueError("value not in array")

    return binary_search(search_list, 0, len(search_list) - 1, value)

File: solutions/fortran/binary-search/1/binary_search.f90
module binary_search
  implicit none
contains

  ! This function performs binary search on a sorted array.
  function find(array, val) result(idx)
    ! Input: 
    ! array - A sorted array of integers
    ! val - The integer value to search for in the array
    ! Output: 
    ! idx - The index of the found element, or -1 if not found

    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx
    integer :: start_index, end_index, middle_index

    ! Initialize start and end indices
    start_index = 1
    end_index = size(array)

    ! Perform binary search
    do while (end_index >= start_index)
      ! Calculate the middle index
      middle_index = (end_index + start_index) / 2

      ! Check if the middle element is the one we're searching for
      if (array(middle_index) == val) then
        idx = middle_index
        return
      end if

      ! Adjust search range based on comparison
      if (array(middle_index) > val) then
        end_index = middle_index - 1
      else
        start_index = middle_index + 1
      end if
    end do

    ! If the value is not found, return -1
    idx = -1
  end function find

end module binary_search
